### ISSUE TITLE
Framework: Create Build for ETI OFF

### ISSUE THREAD

**Issue by ZUUL42** [trilinos/Trilinos#5729]
## Enhancement
@trilinos/framework

Create build that sets disables explicit instantiation (ETI).
This may first be tried in dev->master.

**Comment by mhoemmen** [comment]
btw Sierra now enables ETI by default (as of Tue 13 Aug 2019), though this would still be helpful as long as we permit users to disable ETI.

**Comment by jwillenbring** [comment]
With changing the Trilinos default to ETI ON, we will need to add a build with ETI off, or we will lose the ability to build with ETI off. @kddevin mentioned this recently as well. I suggest one of the dev->master builds. I propose the Clang 9 build, but I am open to any other suggestions.

**Comment by mhoemmen** [comment]
Best practice with an ETI OFF test would be to enable "weird" types, just to make sure that downstream dependency inversion for factories is working correctly.

**Comment by ZUUL42** [comment]
@mhoemmen what are you referring to exactly with "weird types"?

**Comment by ZUUL42** [comment]
I've tested a few things with ETI OFF using GCC 7.2.0 and clang 9.0.0. There are a significant amount of errors for both.
To minimize the number coming from GCC, I turned off WError, but I'm still getting 50+ build errors.
[GCC 7.2.0 ETI OFF CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&display=project&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=63&value1=eti_off_werror_off-test-Trilinos_pullrequest_gcc_7.2.0-5807&field2=buildstamp&compare2=63&value2=Experimental)
[clang 9.0.0 ETI OFF CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&display=project&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=63&value1=eti_off-test-Trilinos_pullrequest_clang_9.0.0-1412&field2=buildstamp&compare2=63&value2=Experimental)

It seems we're primarily running into issues in @trilinos/tempus with a few problems in @trilinos/piro and @trilinos/rol and some odd issue in @trilinos/muelu, @trilinos/shylu, @trilinos/ifpack2, and @trilinos/xpetra. Though there are likely more hiding behind the 50 error limiter.
Could the teams take a look and let me know if there is something I can change in my configuration or if there are changes we're going to need to make in the packages. Thanks

**Comment by jhux2** [comment]
@ZUUL42 I randomly sampled the MueLu errors, which are all in the tests.  They are of the type
```
g++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <https://gcc.gnu.org/bugs/> for instructions.
```

I don't know what could be causing this.  You might try reducing the `make` parallelism, just an idea.

**Comment by mhoemmen** [comment]
@ZUUL42 wrote:
> what are you referring to exactly with "weird types"?

I should have clarified: types that are not enabled by default in ETI=ON builds.  This will test run-time registration of solver factories for the ETI=OFF case.

**Comment by ZUUL42** [comment]
@jhux2, I've trying setting [CMAKE_BUILD_PARALLEL_LEVEL to 2](https://github.com/ZUUL42/Trilinos/blob/93e1b2cabccedfc5d8254234b0d40738d9b2ae32/cmake/std/PullRequestLinuxClang9.0.0TestingSettings.cmake#L9) with no change in results.
[Build 1468 w/Tempus OFF](https://testing.sandia.gov/cdash/index.php?project=Trilinos&display=project&filtercount=3&showfilters=1&filtercombine=and&field1=site&compare1=61&value1=ascic141&field2=buildname&compare2=61&value2=PR-eti_off_TempusOFF-test-Trilinos_pullrequest_clang_9.0.0-1468&field3=buildstamp&compare3=61&value3=20210203-1100-Experimental)
[Build 1471 w/Tempus OFF & parallel = 2](https://testing.sandia.gov/cdash/index.php?project=Trilinos&display=project&filtercount=3&showfilters=1&filtercombine=and&field1=site&compare1=61&value1=ascic141&field2=buildname&compare2=61&value2=PR-eti_off_parallel2-test-Trilinos_pullrequest_clang_9.0.0-1471&field3=buildstamp&compare3=61&value3=20210203-1842-Experimental)

**Comment by ccober6** [comment]
FYI, the Tempus failures should be fixed in the next day or so with #8549.  Just waiting on a review.

**Comment by ccober6** [comment]
#8549 has been merged, so you should not see any Tempus failures for ETI=OFF.

**Comment by ZUUL42** [comment]
I've run a build of ETI Off with the Master Merge PR Clang 9.0.0 build.
Here are the results: [CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&display=project&filtercount=3&showfilters=1&filtercombine=and&field1=site&compare1=61&value1=ascic141&field2=buildname&compare2=61&value2=PR-eti_off-test-Trilinos_pullrequest_clang_9.0.0-1589&field3=buildstamp&compare3=61&value3=20210524-1524-Experimental)
Unfortunately it hit the 50 error limiter, but we can see build error counts for:
- @trilinos/muelu - 24
- @trilinos/zoltan2 - 13
- @trilinos/stk - 6
- @trilinos/shylu _DD - 2
- @trilinos/ifpack2 - 1
- @trilinos/panzer - 1
- @trilinos/piro - 1
- @trilinos/trilinoscouplings - 1
- @trilinos/xpetra - 1
- (other build errors that didn't appear in the CDash report)

**Comment by cgcgcg** [comment]
@ZUUL42 MueLu build errors should be fixed.

**Comment by ZUUL42** [comment]
Query to show all eti_off test builds. [CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&display=project&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=63&value1=eti_off-test-Trilinos_pullrequest_clang_9.0.0&field2=buildstamp&compare2=63&value2=Experimental)
(I'll edit when it completes.)

**Comment by cgcgcg** [comment]
Looks like some issue in KokkosKernels is now causing a lot of stuff in MueLu to fail?

**Comment by brian-kelley** [comment]
@cgcgcg Something strange is going on with name lookup. I replicated this, the Distance 2 coloring error is:
```
//in included file
namespace KokkosGraph {
namespace Impl {
  class GraphColorDistance2 {...};
}}
//in main D2 coloring header
namespace KokkosGraph {
namespace Experimental {
  void graph_color_distance2(...) {
    Impl::GraphColorDistance2(...);
  }
}}
```
Since ``graph_color_distance2()`` is inside namespace KokkosGraph, I don't know why it can't find Impl::GraphColorDistance2. This works fine in builds with ETI, but D2 coloring doesn't even use ETI. There are other errors with name lookup in MueLu, like:
```
/ascldap/users/bmkelle/MiscWork/Trilinos/packages/muelu/src/Interface/../MueCentral/MueLu_FactoryManager_def.hpp:201:108:
error: expected type-specifier before ‘InitialBlockNumberFactory’
201 |       if (varName == "BlockNumber")                     return SetAndReturnDefaultFactory(varName, rcp(new InitialBlockNumberFactory()));
```

But just replacing the 3 instances of ``Impl::GraphColorDistance2`` with ``KokkosGraph::Impl::GraphColorDistance2`` in kokkos-kernels/src/graph/KokkosGraph_Distance2Color.hpp fixes those errors.

**Comment by cgcgcg** [comment]
Yup, that's rather odd indeed. @brian-kelley do you want to make that change to KokkosKernels, or rather investigate?

**Comment by brian-kelley** [comment]
@cgcgcg I can make the KokkosKernels changes in Trilinos after the 3.4.1 promotion is merged, which should happen today.

**Comment by ZUUL42** [comment]
Ran another test build with ETI = OFF.
[CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&display=project&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=PR-eti_off-test-Trilinos_pullrequest_clang_9.0.0-1594&field2=buildstamp&compare2=61&value2=20210601-2033-Experimental)

**Comment by ikalash** [comment]
Coincidentally I tried doing a Trilinos build with ETI off last week and there were failures in Kokkos and Muelu IIRC.  It would be great to get Trilinos to work w/o ETI, even though this may not be an officially supported feature anymore.  Having an ETI-free build is currently needed for some additions I'm making to the Alegra code.  The ETI-free build was working with Trilinos 13.0 which is what I am using for now.

**Comment by ccober6** [comment]
@trilinos/ifpack2 @trilinos/muelu @trilinos/piro @trilinos/shylu @trilinos/xpetra @jwillenbring (for TrilinosCouplings)  Can you look over your packages errors?  Thanks, Curt

**Comment by mperego** [comment]
@ccober6 Piro errors seem to be a consequence of MueLu errors. I'll check again after MueLu tests have been fixed.

**Comment by cgcgcg** [comment]
I'll have a look at the MueLu errors.

**Comment by srajama1** [comment]
@ccober6 The shylu errors are basically MueLu errors, will revisit after MueLu is done.

**Comment by cgcgcg** [comment]
@ZUUL42 Could you kick off another round of testing? Hopefully I got all the MueLu errors this time.

**Comment by ZUUL42** [comment]
@cgcgcg 
Building with ETI OFF for the latest update to develop. [CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&display=project&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=63&value1=eti_off-test-Trilinos_pullrequest_clang_9.0.0-1599&field2=buildstamp&compare2=63&value2=Experimental)

**Comment by jhux2** [comment]
@ZUUL42 These cmake warnings look a bit scary:

```

Total time to configure Trilinos: 1m16.166s
-- Configuring done
CMake Warning:
  Value of Trilinos_ENABLE_TESTS contained a newline; truncating


CMake Warning:
  Value of Trilinos_PARALLEL_LINK_JOBS_LIMIT contained a newline; truncating


-- Generating done
CMake Warning:
  Value of Trilinos_ENABLE_TESTS contained a newline; truncating


CMake Warning:
  Value of Trilinos_PARALLEL_LINK_JOBS_LIMIT contained a newline; truncating
```

**Comment by ZUUL42** [comment]
@jhux2, I'll look into that.
Though they seem to have been there without this change at least as far back as the [CDash reports on record](https://testing.sandia.gov/cdash/index.php?project=Trilinos&display=project&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=63&value1=test-Trilinos_pullrequest_clang_9.0.0&field2=buildstarttime&compare2=84&value2=NOW), 2021 Feb 04.

**Comment by cgcgcg** [comment]
@ZUUL42 I fixed one of the 2 build errors. The other one looks like the OOM killer might have gotten the compiler. On my workstation, I had to reduce the build parallelism to avoid the issue.

**Comment by ikalash** [comment]
FYI, it appears there are new errors from Ifpack2 if tests are enabled - it's due to the unit tests.  I opened a separate issue on this #9280 .  These just cropped up end of last week with an Ifpack2 PR that went in.

**Comment by ZUUL42** [comment]
I ran another test build the other day, forgot to post right away. (whoops, sorry) This is from Jun 18.
[CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&display=project&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=PR-eti_off-test-Trilinos_pullrequest_clang_9.0.0-1615&field2=buildstamp&compare2=63&value2=Experimental)
I can kick off another as well.

**Comment by cgcgcg** [comment]
Looks like OOM, no?

**Comment by ZUUL42** [comment]
@cgcgcg, that's what I'm finding. 
I've tried setting parallel_level to 8 and then 2 but still getting the same errors.
Might you have any other suggestions?
Otherwise, it seems we'll have to turn off packages or some other creative solution that won't make for a very good PR build.

**Comment by jhux2** [comment]
@ZUUL42 I assume `parallel_level` is akin to `make -j`.  Here a couple ideas to try:

1) Try gcc instead of clang to see if the former is better with memory.
2) Try building with "-O0" and without "-g".  Perhaps this would need less memory then the current `RELEASE` build.

**Comment by jwillenbring** [comment]
I was unable to see the latest CDash results. @ZUUL42 do you remember how many packages were failing? Could we turn a few packages off and get a clean build? There is renewed concern about this.

**Comment by ZUUL42** [comment]
@jwillenbring here's a couple new ETI OFF builds.
The clang 9 build still has parallel level set to 1. The clang 10 build has no other changes besides having ETI set to off.
[CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&display=project&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=63&value1=eti_off&field2=buildstamp&compare2=63&value2=Experimental)

**Comment by ZUUL42** [comment]
Kicked off a couple more builds. The same link will pick them up.
The Clang 10 build has the 34 failing tests turned off.
The Clang 9 build, so far, just has the parallel level line removed.

**Comment by jhux2** [comment]
@ZUUL42 You might try a different compiler, such as gcc, which presumably would have different memory requirements (if memory is what's causing the internal compiler error).

**Comment by ZUUL42** [comment]
@jhux2 I _can_ do that. Though I'm not sure if there was a reason Clang was specifically chosen for the ETI OFF build or not.
@jwillenbring it appears you had proposed clang, at least in this thread. Do you recall if there was a specific reason for that?
The test results are gone, but it appears I had run ETI OFF with clang 9 & gcc 7.2. I seem to remember clang 9 appeared to be closer to building clean, so I went that direction. However there have been a significant amount of change since then.
I can run it through with GCC 7.2 & 8.3 and see what we get.

**Comment by jhux2** [comment]
@ZUUL42 I'm just thinking of how to avoid what appears to be an out-of-memory error with the compiler.

The builds are all `RELEASE`.  A clang build with `-O0` and without `-g` might relieve some of the memory pressure and allow you to reenable those MueLu tests.

**Comment by cgcgcg** [comment]
The MueLu tests were failing due to a bug in Belos, not due to OOM, no?

**Comment by jhux2** [comment]
> The MueLu tests were failing due to a bug in Belos, not due to OOM, no?

@cgcgcg The results in @ZUUL42's [comment](https://github.com/trilinos/Trilinos/issues/5729#issuecomment-890295226) still have internal compiler errors for MueLu, which I interpret as OOM.

**Comment by ZUUL42** [comment]
@jhux2 I'm not familiar with those flags.
For CMake, would it be:
`set(CMAKE_CXX_FLAGS "-O0 -g" CACHE STRING)`
in [PullRequestLinuxClang10.0.0TestingSettings.cmake](https://github.com/ZUUL42/Trilinos/blob/eti_off/cmake/std/PullRequestLinuxClang10.0.0TestingSettings.cmake)

**Comment by jhux2** [comment]
@ZUUL42 You probably only want `-O0`, *without* the `-g` option.  `O0` sets the compiler optimization level to zero, i.e., no optimization' .  Adding `-g` would cause the compiler to include symbols and thus increase the executable size.

**Comment by cgcgcg** [comment]
Looks like debug symbols are also switched on here:
https://github.com/trilinos/Trilinos/blob/a5b9f3dec9940fa684c70aa6bf1fc89c97fe4150/cmake/std/PullRequestLinuxCommonTestingSettings.cmake#L30

**Comment by ZUUL42** [comment]
Clang 10 Build #3272 adds:
`set(CMAKE_CXX_FLAGS_RELEASE "-O0")`
and removes:
`set (Trilinos_ENABLE_DEBUG_SYMBOLS ON CACHE BOOL "Set by default for PR testing")`
With tests back on ... 35 of them still fail. See [CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&display=project&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=63&value1=eti_off&field2=buildstamp&compare2=63&value2=Experimental&field3=configureerrors&compare3=41&value3=0)

**Comment by jhux2** [comment]
> Clang 10 Build 3272 adds:
> `set(CMAKE_CXX_FLAGS_RELEASE "-O0")`
> and removes:
> `set (Trilinos_ENABLE_DEBUG_SYMBOLS ON CACHE BOOL "Set by default for PR testing")`
> With tests back on ... 35 of them still fail. See [CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&display=project&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=63&value1=eti_off&field2=buildstamp&compare2=63&value2=Experimental&field3=configureerrors&compare3=41&value3=0)

@ZUUL42 This is progress, right?  The MueLu tests that previously didn't compile now do, right?

**Comment by wadeburgess** [comment]
Migrated to Jira.
